Sources/xcproj/Referenceable.swift

```diff
@@ -0,0 +1,112 @@
+//
+//  Referenceable.swift
+//  xcprojPackageDescription
+//
+//  Created by rmalik on 11/8/17.
+//
+
+import Foundation
+
+// Typealias that represents dictionary from string -> T where T is Referenceable (i.e. PBXObjects)
+
+public typealias ReferenceableCache<T: Referenceable & Equatable> = [String: T]
+
+public protocol Referenceable {
+    var reference: String { get }
+}
+
+extension Array where Element: Referenceable {
+
+    public var references: [String] {
+        return map { $0.reference }
+    }
+
+    public func contains(reference: String) -> Bool {
+        return contains { $0.reference == reference }
+    }
+
+    public func getReference(_ reference: String) -> Element? {
+        return first { $0.reference == reference }
+    }
+}
+
+
+extension Dictionary where Key == String, Value: Referenceable {
+    static func from(references: [Value]) -> Dictionary<String, Value> {
+        return Dictionary<String, Value>(uniqueKeysWithValues: references.flatMap { ($0.reference, $0) })
+    }
+
+    public var references: [String] {
+        return self.values.map { $0.reference }
```

Comment:
Isn't this just self.keys? This again brings up the possible conflict between the key of the dictionary and the reference property. The reference property should be made a let.

The recent idea of removing the reference property entirely is an interesting one too. The use cases would be to be thought about, but maybe that's a good move